### PR TITLE
[Merged by Bors] - Implement liveness BeaconAPI

### PR DESCRIPTION
**Pull request by aoikurokawa**
## Issue Addressed

#4243

## Proposed Changes

- create a new endpoint for liveness/{endpoint}

## Additional Info
This is my first PR.

### CODE REVIEW

File: beacon_node/http_api/src/lib.rs [hunk 1]
```diff
@@ -3246,6 +3247,46 @@ pub fn serve<T: BeaconChainTypes>(
             },
         );
 
+    // POST lighthouse/liveness/{epoch}
```
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: The path should be `/eth/v1/validator/liveness/{epoch}`, matching the spec: https://ethereum.github.io/beacon-APIs/#/Validator/getLiveness

File: beacon_node/http_api/src/lib.rs [hunk 2]
```diff
@@ -3246,6 +3247,46 @@ pub fn serve<T: BeaconChainTypes>(
...
+        .and_then(
+            |epoch: Epoch, indices: Vec<u64>, chain: Arc<BeaconChain<T>>| {
+                blocking_json_task(move || {
+                    // Ensure the request is for either the current, previous or next epoch.
+                    let current_epoch = chain
+                        .epoch()
+                        .map_err(warp_utils::reject::beacon_chain_error)?;
+                    let next_epoch = current_epoch.saturating_add(Epoch::new(1));
+
+                    if epoch > next_epoch {
```
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: The prev_epoch check is missing

File: beacon_node/http_api/src/lib.rs [hunk 3]
```diff
@@ -3246,6 +3247,46 @@ pub fn serve<T: BeaconChainTypes>(
...
+                    }
+
+                    let liveness: Vec<api_types::LivenessResponseData> = indices
+                        .iter()
+                        .cloned()
+                        .map(|index| {
+                            let is_live = chain.validator_seen_at_epoch(index as usize, epoch);
+                            api_types::LivenessResponseData {
+                                index,
+                                epoch,
```
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: We need a new type here, like `StandardLivenessResponseData` which does not include the `epoch`. The `epoch` is not part of the standardised endpoint.

File: beacon_node/http_api/src/lib.rs [hunk 4]
```diff
@@ -3864,6 +3905,7 @@ pub fn serve<T: BeaconChainTypes>(
                     .uor(post_validator_prepare_beacon_proposer)
                     .uor(post_validator_register_validator)
                     .uor(post_lighthouse_liveness)
+                    .uor(post_lighthouse_liveness_epoch)
```
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: this should be renamed `post_validator_liveness` and moved up with the other `post_validator_*` methods

File: common/eth2/src/lib.rs [hunk 5]
```diff
@@ -1543,6 +1543,28 @@ impl BeaconNodeHttpClient {
         .await
     }
 
+    /// `POST lighthouse/liveness/{epoch}`
+    pub async fn post_lighthouse_liveness_epoch(
```
Comment (michaelsproul) [inline common/eth2/src/lib.rs]:
```suggestion
    pub async fn post_validator_liveness_epoch(
```

Similarly this should be renamed and refactored to use the `/eth/v1/validator/liveness` path

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Thank you for the contribution!

The code is very well-written overall, I just had a few suggestions about how to achieve conformance with the spec.

Thanks!

**Comment by aoikurokawa** [comment]
Thank you for reviewing!. 

I think I have fixed all what you reviewed for my code.

**Comment by paulhauner** [comment]
@aoikurokawa the `cargo-fmt` check is failing. You can fix this my running `cargo fmt` in the project root and committing the result ☺️

**Comment by aoikurokawa** [comment]
Oops, sorry. 
I run the command `cargo fmt`, then committed, 
I appreciate your review.

**Review by jimmygchen — APPROVED** [review]
Looks good to me 👍 

There's just one `clippy` failure that I think should be fixed by merging latest `unstable` branch into this branch, and we should be good to merge!

**Comment by aoikurokawa** [comment]
@jimmygchen  Thank you for reviewing 🙏
I think I have merged the current unstable branch.

**Review by paulhauner — APPROVED** [review]
Very neat and tidy!

**Pull request closed without merging**